File: src/automation/AutomationEngine.ts
import {
  Action,
  ActionType,
  ExecutionState,
  ActionResult,
  VerificationResult,
  AutomationPlan,
  AutomationDefinition,
  AutomationLogEntry,
  RiskLevel,
} from './types';
import { ActionValidator } from './ActionValidator';
import { AppControlExecutor } from './AppControlExecutor';
import { UIActionExecutor } from './UIActionExecutor';
import { GestureExecutor } from './GestureExecutor';
import { TextExecutor } from './TextExecutor';
import { MediaControlExecutor } from './MediaControlExecutor';
import { SystemAutomationExecutor } from './SystemAutomationExecutor';
import { CommunicationExecutor } from './CommunicationExecutor';
import { notificationAutomationService } from './NotificationAutomationService';
import { ruleEngine } from './RuleEngine';
import { SafetyValidator } from './SafetyValidator';
import { confirmationManager } from './ConfirmationManager';
import { LoopProtection } from './LoopProtection';
import { automationLogService } from './AutomationLogService';
import { automationRepository } from './AutomationRepository';
import { AutomationAccessibilityService } from './AutomationAccessibilityService';
import { AutomationPermissionManager } from './AutomationPermissionManager';

export interface AutomationEngineState {
  engineState: ExecutionState;
  currentPlanTitle: string;
  currentActionIndex: number;
  totalActions: number;
  currentAction: Action | null;
  history: Action[];
  queue: Action[];
  isEmergencyStopped: boolean;
  activeError: string | null;
  lastCompletedAt: number | null;
}

export class AutomationEngine {
  private static instance: AutomationEngine;

  private state: AutomationEngineState = {
    engineState: ExecutionState.IDLE,
    currentPlanTitle: '',
    currentActionIndex: -1,
    totalActions: 0,
    currentAction: null,
    history: [],
    queue: [],
    isEmergencyStopped: false,
    activeError: null,
    lastCompletedAt: null,
  };

  private abortController: AbortController | null = null;
  private stateListeners: ((state: AutomationEngineState) => void)[] = [];
  private permissionManager = AutomationPermissionManager.getInstance();
  private accessibilityService = AutomationAccessibilityService.getInstance();

  public static getInstance(): AutomationEngine {
    if (!AutomationEngine.instance) {
      AutomationEngine.instance = new AutomationEngine();
    }
    return AutomationEngine.instance;
  }

  public getState(): AutomationEngineState {
    return { ...this.state };
  }

  public onStateChange(listener: (state: AutomationEngineState) => void): () => void {
    this.stateListeners.push(listener);
    listener(this.getState());
    return () => {
      this.stateListeners = this.stateListeners.filter((l) => l !== listener);
    };
  }

  private notify(): void {
    const s = this.getState();
    this.stateListeners.forEach((l) => l(s));
  }

  /**
   * EMERGENCY STOP: Immediately cancels all currently running and queued actions.
   * Can be invoked via voice/text "STOP AUTOMATION" or explicit red UI button.
   */
  public emergencyStop(reason: string = 'User triggered emergency stop'): void {
    this.state.isEmergencyStopped = true;
    this.state.activeError = reason;
    this.state.engineState = ExecutionState.CANCELLED;

    if (this.abortController) {
      this.abortController.abort(reason);
      this.abortController = null;
    }

    if (this.state.currentAction) {
      this.state.currentAction.executionState = ExecutionState.CANCELLED;
      this.state.currentAction.error = reason;
      this.state.history.push({ ...this.state.currentAction });
      this.state.currentAction = null;
    }

    // Drain remaining queue as cancelled
    while (this.state.queue.length > 0) {
      const remaining = this.state.queue.shift()!;
      remaining.executionState = ExecutionState.CANCELLED;
      remaining.error = reason;
      this.state.history.push(remaining);
    }

    this.notify();
  }

  /**
   * Resets the emergency stop latch to allow new actions.
   */
  public resetEmergencyStop(): void {
    this.state.isEmergencyStopped = false;
    this.state.activeError = null;
    this.state.engineState = ExecutionState.IDLE;
    this.notify();
  }

  /**
   * Executes a structured automation plan sequentially with verification, retries, and timeouts.
   */
  public async executePlan(
    title: string,
    rawActions: Partial<Action>[],
    sensitiveAppsDenylist: string[] = []
  ): Promise<{ success: boolean; executedCount: number; message: string }> {
    if (this.state.isEmergencyStopped) {
      return {
        success: false,
        executedCount: 0,
        message: 'Engine is in Emergency Stopped state. Reset required.',
      };
    }

    this.state.engineState = ExecutionState.PLANNING;
    this.state.currentPlanTitle = title;
    this.state.activeError = null;
    this.notify();

    // 1. Validation Phase
    this.state.engineState = ExecutionState.VALIDATING;
    this.notify();

    const validatedActions: Action[] = [];
    for (const raw of rawActions) {
      const validation = ActionValidator.validate(raw, sensitiveAppsDenylist);
      if (!validation.valid || !validation.sanitizedAction) {
        this.state.engineState = ExecutionState.FAILED;
        this.state.activeError = `Validation failed: ${validation.errors.join('; ')}`;
        this.notify();
        return {
          success: false,
          executedCount: 0,
          message: this.state.activeError,
        };
      }
      validatedActions.push(validation.sanitizedAction);
    }

    // Sort by priority if applicable
    validatedActions.sort((a, b) => (b.priority ?? 1) - (a.priority ?? 1));

    this.state.queue = [...validatedActions];
    this.state.totalActions = validatedActions.length;
    this.state.currentActionIndex = 0;
    this.abortController = new AbortController();

    let completedCount = 0;

    // 2. Sequential Execution Loop
    for (let i = 0; i < validatedActions.length; i++) {
      if (this.state.isEmergencyStopped || !this.abortController || this.abortController.signal?.aborted) {
        this.state.engineState = ExecutionState.CANCELLED;
        this.notify();
        return {
          success: false,
          executedCount: completedCount,
          message: 'Automation cancelled.',
        };
      }

      const action = this.state.queue.shift()!;
      this.state.currentAction = action;
      this.state.currentActionIndex = i;

      // Permission Check
      const permCheck = this.permissionManager.checkPermissionsForAction(action.requiredPermissions);
      if (!permCheck.allGranted) {
        this.state.engineState = ExecutionState.WAITING_PERMISSION;
        action.executionState = ExecutionState.WAITING_PERMISSION;
        action.error = `Missing permissions: ${permCheck.missing.join(', ')}`;
        this.notify();
        return {
          success: false,
          executedCount: completedCount,
          message: `Permission denied: ${permCheck.missing.join(', ')}`,
        };
      }

      // Execute Single Action with Timeout and Retries
      const executionResult = await this.executeActionWithRetry(action, sensitiveAppsDenylist);

      action.result = executionResult.result;
      action.verificationResult = executionResult.verification;
      action.executionState = executionResult.finalState;
      action.error = executionResult.error;

      this.state.history.push({ ...action });

      if (executionResult.finalState === ExecutionState.COMPLETED) {
        completedCount++;
      } else {
        this.state.engineState = executionResult.finalState;
        this.state.activeError = executionResult.error || 'Action failed.';
        this.notify();
        return {
          success: false,
          executedCount: completedCount,
          message: executionResult.error || 'Action failed.',
        };
      }
    }

    this.state.engineState = ExecutionState.COMPLETED;
    this.state.currentAction = null;
    this.state.lastCompletedAt = Date.now();
    this.notify();

    return {
      success: true,
      executedCount: completedCount,
      message: `Successfully completed ${completedCount} actions for "${title}".`,
    };
  }

  /**
   * Executes an action with timeout race and failure retry handling (reinspects UI on retry).
   */
  private async executeActionWithRetry(
    action: Action,
    sensitiveDenylist: string[]
  ): Promise<{
    finalState: ExecutionState;
    result?: ActionResult;
    verification?: VerificationResult;
    error?: string;
  }> {
    let attempts = 0;
    const maxRetries = action.retryCount ?? 2;

    while (attempts <= maxRetries) {
      if (this.state.isEmergencyStopped || !this.abortController || this.abortController.signal?.aborted) {
        return { finalState: ExecutionState.CANCELLED, error: 'Cancelled by emergency stop' };
      }

      if (attempts > 0) {
        this.state.engineState = ExecutionState.RETRYING;
        action.executionState = ExecutionState.RETRYING;
        action.retriesAttempted = attempts;
        this.notify();
        // Failure recovery step: wait and reinspect UI
        await new Promise((r) => setTimeout(r, 500));
        this.accessibilityService.getRootInActiveWindow();
      }

      this.state.engineState = ExecutionState.EXECUTING;
      action.executionState = ExecutionState.EXECUTING;
      this.notify();

      try {
        // Enforce Timeout with Promise.race
        const timeoutPromise = new Promise<{ isTimeout: true }>((_, reject) =>
          setTimeout(() => reject(new Error('TIMEOUT')), action.timeout)
        );

        const executionPromise = this.dispatchToExecutor(action, sensitiveDenylist);
        const { result, verification } = await Promise.race([executionPromise, timeoutPromise]) as any;

        // 3. Verification Phase
        this.state.engineState = ExecutionState.VERIFYING;
        action.executionState = ExecutionState.VERIFYING;
        this.notify();

        if (result.success && verification?.verified) {
          return {
            finalState: ExecutionState.COMPLETED,
            result,
            verification,
          };
        }

        // Action or verification failed, increment attempt
        attempts++;
        if (attempts > maxRetries) {
          return {
            finalState: ExecutionState.FAILED,
            result,
            verification,
            error: result.message || verification?.failureReason || 'Action failed verification after max retries.',
          };
        }
      } catch (err: any) {
        if (err.message === 'TIMEOUT') {
          return {
            finalState: ExecutionState.TIMEOUT,
            error: `Action exceeded timeout limit of ${action.timeout}ms.`,
          };
        }

        attempts++;
        if (attempts > maxRetries) {
          return {
            finalState: ExecutionState.FAILED,
            error: err.message || 'Execution failed unexpectedly.',
          };
        }
      }
    }

    return {
      finalState: ExecutionState.FAILED,
      error: 'Max retries exhausted.',
    };
  }

  /**
   * Dispatches an action to its specific specialized Android executor.
   */
  private async dispatchToExecutor(
    action: Action,
    sensitiveDenylist: string[]
  ): Promise<{ result: ActionResult; verification: VerificationResult }> {
    const params = action.parameters || {};

    switch (action.actionType) {
      case ActionType.APP_ACTION:
        if (params.operation === 'deep_link' && params.uri) {
          return await AppControlExecutor.openDeepLink(params.uri);
        }
        if (params.operation === 'send_intent') {
          return await AppControlExecutor.sendIntent(params.mimeType || 'text/plain', params.content || '', params.targetPkg);
        }
        if (params.operation === 'system_settings' && params.settingPanel) {
          return await AppControlExecutor.openSystemSetting(params.settingPanel);
        }
        if (params.operation === 'back') {
          return await AppControlExecutor.navigateBack();
        }
        return await AppControlExecutor.launchApp(params.appName || params.packageName, sensitiveDenylist);

      case ActionType.UI_ACTION:
        const query = {
          resourceId: params.resourceId,
          contentDescription: params.contentDescription,
          text: params.text,
          isClickable: params.isClickable,
          isEditable: params.isEditable,
          structuralIndex: params.structuralIndex,
        };

        if (params.operation === 'long_click') {
          return await UIActionExecutor.longClick(query);
        }
        if (params.operation === 'focus') {
          return await UIActionExecutor.focus(query);
        }
        if (params.operation === 'scroll') {
          return await UIActionExecutor.scroll(params.direction || 'forward');
        }
        if (params.operation === 'back') {
          return await AppControlExecutor.navigateBack();
        }
        if (params.operation === 'wait') {
          const waitRes = await UIActionExecutor.waitForElement(query, params.timeoutMs || 5000);
          return {
            result: {
              success: waitRes.verified,
              message: waitRes.verified ? 'Target element appeared.' : 'Wait for element timed out.',
              executionTimeMs: waitRes.waitTimeMs,
            },
            verification: {
              verified: waitRes.verified,
              observedState: waitRes.verified ? 'ELEMENT_APPEARED' : 'ELEMENT_NOT_APPEARED',
            },
          };
        }
        return await UIActionExecutor.click(query);

      case ActionType.GESTURE_ACTION:
        return await GestureExecutor.executeGesture(params.gesture, params.coords);

      case ActionType.TEXT_ACTION:
        const tQuery = {
          resourceId: params.resourceId,
          text: params.targetText,
          isEditable: true,
        };
        if (params.operation === 'submit') {
          return await TextExecutor.submitText(tQuery);
        }
        return await TextExecutor.enterText(tQuery, params.text || '', params.operation || 'enter');

      case ActionType.DELAY_ACTION:
        const duration = Number(params.durationMs) || 1000;
        await new Promise((r) => setTimeout(r, duration));
        return {
          result: {
            success: true,
            message: `Waited for ${duration}ms delay.`,
            executionTimeMs: duration,
          },
          verification: {
            verified: true,
            observedState: 'DELAY_COMPLETED',
          },
        };

      case ActionType.MEDIA_ACTION:
        if (params.command === 'set_volume') {
          return await MediaControlExecutor.setMediaVolume(params.volumePercent ?? 50);
        }
        return await MediaControlExecutor.executeMediaCommand(params.command || 'play', params.targetApp);

      case ActionType.SYSTEM_ACTION:
        if (params.operation === 'brightness') {
          return await SystemAutomationExecutor.setBrightness(params.brightnessPercent ?? 50);
        }
        if (params.operation === 'settings') {
          return await SystemAutomationExecutor.openSettingsPanel(params.panel || 'wifi');
        }
        return await SystemAutomationExecutor.setSystemVolume(params.streamType || 'media', params.volumePercent ?? 50);

      case ActionType.COMMUNICATION_ACTION:
        if (params.operation === 'dial') {
          return await CommunicationExecutor.openDialer(params.phoneNumber || '');
        }
        if (params.operation === 'call') {
          return await CommunicationExecutor.makePhoneCall(params.phoneNumber || params.contactName || '');
        }
        if (params.operation === 'sms') {
          return await CommunicationExecutor.composeSms(params.phoneNumber || '', params.messageBody || '');
        }
        if (params.operation === 'share') {
          return await CommunicationExecutor.shareContent(params.text || '', params.title);
        }
        return await CommunicationExecutor.openDialer(params.phoneNumber || '');

      case ActionType.NOTIFICATION_ACTION:
        if (params.operation === 'dismiss' && params.notificationId) {
          const dismissed = notificationAutomationService.dismissNotification(params.notificationId);
          return {
            result: {
              success: dismissed,
              message: dismissed ? 'Notification dismissed.' : 'Notification not found.',
              executionTimeMs: 40,
            },
            verification: {
              verified: dismissed,
              observedState: dismissed ? 'NOTIF_DISMISSED' : 'NOTIF_FAILED',
            },
          };
        }
        if (params.operation === 'open' && params.notificationId) {
          const opened = await notificationAutomationService.openNotificationApp(params.notificationId);
          return {
            result: {
              success: opened,
              message: opened ? 'Notification app opened.' : 'Failed to open notification app.',
              executionTimeMs: 150,
            },
            verification: {
              verified: opened,
              observedState: opened ? 'NOTIF_APP_OPENED' : 'NOTIF_APP_FAILED',
            },
          };
        }
        return {
          result: { success: true, message: 'Notification action processed', executionTimeMs: 20 },
          verification: { verified: true, observedState: 'NOTIF_PROCESSED' },
        };

      default:
        return {
          result: {
            success: true,
            message: `Dispatched action ${action.actionType}`,
            executionTimeMs: 50,
          },
          verification: {
            verified: true,
            observedState: 'DISPATCHED',
          },
        };
    }
  }

  /**
   * Executes a saved or newly configured AutomationDefinition with preconditions, safety check,
   * loop protection, verification rules, and execution logging.
   */
  public async executeAutomationDefinition(
    def: AutomationDefinition,
    options: { skipConfirmation?: boolean } = {}
  ): Promise<{ success: boolean; executedCount: number; message: string }> {
    const loopProtection = new LoopProtection(def.actions.length + 10, def.timeoutMs || 60000);
    const recursionCheck = loopProtection.checkRecursion(def.id);
    if (!recursionCheck.safe) {
      return { success: false, executedCount: 0, message: recursionCheck.reason! };
    }

    // 1. Evaluate Conditions via RuleEngine
    if (def.conditions && def.conditions.length > 0) {
      const condEval = ruleEngine.evaluateConditions(def.conditions);
      if (!condEval.satisfied) {
        return {
          success: false,
          executedCount: 0,
          message: `Preconditions not met: ${condEval.summary}`,
        };
      }
    }

    // 2. Safety & Confirmation Gate
    const safetyCheck = SafetyValidator.validatePlanSafety(def.actions, options.skipConfirmation, def.confirmationPolicy);
    if (safetyCheck.requiresUserConfirmation) {
      const firstSensitive = def.actions.find((a) => SafetyValidator.classifyActionRisk(a) !== RiskLevel.LOW_RISK) || def.actions[0];
      const approved = await confirmationManager.requestConfirmation(
        def.name,
        firstSensitive,
        safetyCheck.riskLevel,
        safetyCheck.reasons.join(' ') || 'This automation contains actions requiring user consent.'
      );
      if (!approved) {
        return {
          success: false,
          executedCount: 0,
          message: 'Automation execution cancelled by user: Confirmation denied.',
        };
      }
    }

    // 3. Execute Actions
    const startTime = Date.now();
    const planResult = await this.executePlan(def.name, def.actions);
    const duration = Date.now() - startTime;

    // 4. Record to Log Service & Repository
    const logEntry: AutomationLogEntry = {
      id: `log_${Date.now()}_${Math.random().toString(36).substring(2, 6)}`,
      automationId: def.id,
      automationName: def.name,
      startTime,
      endTime: Date.now(),
      durationMs: duration,
      totalActions: def.actions.length,
      successfulActions: planResult.executedCount,
      failedActions: planResult.success ? 0 : 1,
      failureReason: planResult.success ? undefined : planResult.message,
      finalStatus: planResult.success ? 'SUCCESS' : 'FAILED',
      stepDetails: this.state.history.slice(-def.actions.length).map((a) => ({
        actionTitle: a.title,
        status: a.executionState === ExecutionState.COMPLETED ? 'success' : 'failed',
        durationMs: a.result?.executionTimeMs || 0,
        error: a.error,
      })),
    };

    automationLogService.recordLog(logEntry);
    automationRepository.recordExecution(def.id, planResult.success ? 'success' : 'failure');

    return planResult;
  }
}
export const automationEngine = AutomationEngine.getInstance();

File: app/src/main/java/com/nova/assistant/automation/AutomationEngine.kt
package com.nova.assistant.automation

import android.content.Context
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.concurrent.ConcurrentLinkedQueue

data class EngineStatus(
    val state: ExecutionState,
    val planTitle: String = "",
    val activeAction: Action? = null,
    val completedCount: Int = 0,
    val totalCount: Int = 0,
    val isEmergencyStopped: Boolean = false,
    val errorMessage: String? = null
)

/**
 * Production AutomationEngine
 * Coordinates sequential actions, queue management, timeouts, retries,
 * failure recovery, verification, and emergency stops.
 */
class AutomationEngine(private val context: Context) {

    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    private var executionJob: Job? = null

    private val actionQueue = ConcurrentLinkedQueue<Action>()
    private val actionHistory = mutableListOf<Action>()

    private val _engineStatus = MutableStateFlow(EngineStatus(ExecutionState.IDLE))
    val engineStatus: StateFlow<EngineStatus> = _engineStatus.asStateFlow()

    private val appControlExecutor = AppControlExecutor(context)
    private val permissionManager = AutomationPermissionManager(context)

    /**
     * EMERGENCY STOP: Cancels all active and queued actions immediately.
     */
    fun emergencyStop(reason: String = "User requested EMERGENCY STOP") {
        executionJob?.cancel(CancellationException(reason))
        actionQueue.clear()

        _engineStatus.value = _engineStatus.value.copy(
            state = ExecutionState.CANCELLED,
            isEmergencyStopped = true,
            errorMessage = reason
        )
    }

    /**
     * Resets the emergency stop latch.
     */
    fun resetEmergencyStop() {
        _engineStatus.value = EngineStatus(ExecutionState.IDLE)
    }

    /**
     * Executes an automation plan with validation, timeouts, retries, and verification.
     */
    fun executePlan(
        title: String,
        actions: List<Action>,
        onResult: (Boolean, String) -> Unit
    ) {
        if (_engineStatus.value.isEmergencyStopped) {
            onResult(false, "Engine is in Emergency Stop state. Reset required.")
            return
        }

        executionJob?.cancel()
        executionJob = scope.launch {
            _engineStatus.value = EngineStatus(
                state = ExecutionState.PLANNING,
                planTitle = title,
                totalCount = actions.size
            )

            // 1. Validation Phase
            _engineStatus.value = _engineStatus.value.copy(state = ExecutionState.VALIDATING)
            val validatedActions = mutableListOf<Action>()

            for (act in actions) {
                val validation = ActionValidator.validate(act)
                if (!validation.valid || validation.sanitizedAction == null) {
                    val errMsg = "Action validation failed: ${validation.errors.joinToString("; ")}"
                    _engineStatus.value = _engineStatus.value.copy(
                        state = ExecutionState.FAILED,
                        errorMessage = errMsg
                    )
                    onResult(false, errMsg)
                    return@launch
                }
                validatedActions.add(validation.sanitizedAction)
            }

            // Sort by priority descending
            validatedActions.sortByDescending { it.priority }
            actionQueue.clear()
            actionQueue.addAll(validatedActions)

            var completedCount = 0

            // 2. Execution Loop
            while (!actionQueue.isEmpty()) {
                if (!isActive || _engineStatus.value.isEmergencyStopped) {
                    _engineStatus.value = _engineStatus.value.copy(state = ExecutionState.CANCELLED)
                    onResult(false, "Automation cancelled.")
                    return@launch
                }

                val currentAction = actionQueue.poll() ?: break
                _engineStatus.value = _engineStatus.value.copy(
                    state = ExecutionState.EXECUTING,
                    activeAction = currentAction,
                    completedCount = completedCount
                )

                // Permission check
                if (currentAction.requiredPermissions.isNotEmpty()) {
                    val missing = currentAction.requiredPermissions.filter { !permissionManager.isPermissionGranted(it) }
                    if (missing.isNotEmpty()) {
                        _engineStatus.value = _engineStatus.value.copy(
                            state = ExecutionState.WAITING_PERMISSION,
                            errorMessage = "Missing permissions: ${missing.joinToString()}"
                        )
                        onResult(false, "Permission required: ${missing.joinToString()}")
                        return@launch
                    }
                }

                // Execute action with retries and timeout
                val success = executeWithRetry(currentAction)

                if (success) {
                    completedCount++
                    actionHistory.add(currentAction.copy(executionState = ExecutionState.COMPLETED))
                } else {
                    _engineStatus.value = _engineStatus.value.copy(
                        state = ExecutionState.FAILED,
                        errorMessage = currentAction.error ?: "Action execution failed"
                    )
                    onResult(false, currentAction.error ?: "Failed to execute ${currentAction.actionType}")
                    return@launch
                }
            }

            _engineStatus.value = _engineStatus.value.copy(
                state = ExecutionState.COMPLETED,
                activeAction = null,
                completedCount = completedCount
            )
            onResult(true, "Successfully executed $completedCount actions for '$title'")
        }
    }

    private suspend fun executeWithRetry(action: Action): Boolean {
        var attempts = 0
        val maxRetries = action.retryCount

        while (attempts <= maxRetries) {
            if (attempts > 0) {
                _engineStatus.value = _engineStatus.value.copy(state = ExecutionState.RETRYING)
                delay(500L) // UI reinspection backoff
            }

            _engineStatus.value = _engineStatus.value.copy(state = ExecutionState.EXECUTING)

            try {
                // Enforce timeout per action
                val resultPair = withTimeout(action.timeout) {
                    dispatchAction(action)
                }

                // 3. Verification Phase
                _engineStatus.value = _engineStatus.value.copy(state = ExecutionState.VERIFYING)
                action.result = resultPair.first
                action.verificationResult = resultPair.second

                if (resultPair.first.success && resultPair.second.verified) {
                    action.executionState = ExecutionState.COMPLETED
                    return true
                }

                attempts++
            } catch (e: TimeoutCancellationException) {
                action.executionState = ExecutionState.TIMEOUT
                action.error = "Action timed out after ${action.timeout}ms"
                return false
            } catch (e: Exception) {
                attempts++
                action.error = e.message
            }
        }

        action.executionState = ExecutionState.FAILED
        return false
    }

    private suspend fun dispatchAction(action: Action): Pair<ActionResult, VerificationResult> {
        val params = action.parameters
        return when (action.actionType) {
            ActionType.APP_ACTION -> {
                val pkg = params["packageName"] as? String ?: "com.android.settings"
                appControlExecutor.launchApp(pkg)
            }
            ActionType.UI_ACTION -> {
                val query = UIElementQuery(
                    resourceId = params["resourceId"] as? String,
                    contentDescription = params["contentDescription"] as? String,
                    text = params["text"] as? String
                )
                val op = params["operation"] as? String ?: "click"
                when (op) {
                    "long_click" -> UIActionExecutor.longClick(query)
                    "scroll" -> UIActionExecutor.scroll(forward = true)
                    "back" -> UIActionExecutor.back()
                    else -> UIActionExecutor.click(query)
                }
            }
            ActionType.GESTURE_ACTION -> {
                val gesture = params["gesture"] as? String ?: "tap"
                GestureExecutor.executeGesture(gesture)
            }
            ActionType.TEXT_ACTION -> {
                val query = UIElementQuery(
                    resourceId = params["resourceId"] as? String,
                    text = params["targetText"] as? String
                )
                val text = params["text"] as? String ?: ""
                val mode = params["operation"] as? String ?: "enter"
                TextExecutor.enterText(query, text, mode)
            }
            ActionType.DELAY_ACTION -> {
                val ms = (params["durationMs"] as? Number)?.toLong() ?: 1000L
                delay(ms)
                Pair(
                    ActionResult(true, "Delayed ${ms}ms", executionTimeMs = ms),
                    VerificationResult(true, "DELAY_COMPLETED")
                )
            }
            else -> {
                Pair(
                    ActionResult(true, "Executed ${action.actionType}"),
                    VerificationResult(true, "DEFAULT_EXECUTED")
                )
            }
        }
    }
}
